ghcide/src/Development/IDE/Core/Shake.hs

```diff
@@ -1144,117 +1143,78 @@ defineEarlyCutoff' doDiagnostics cmp key file old mode action = do
                     A res
         liftIO $ atomicallyNamed "define - dirtyKeys" $ modifyTVar' dirtyKeys (HSet.delete $ toKey key file)
         return res
+  where
+    -- Highly unsafe helper to compute the version of a file
+    -- without creating a dependency on the GetModificationTime rule
+    -- (and without creating cycles in the build graph).
+    estimateFileVersionUnsafely
+        :: forall k v
+         . IdeRule k v
+        => Values
+        -> k
+        -> Maybe v
+        -> NormalizedFilePath
+        -> Action (Maybe FileVersion)
+    estimateFileVersionUnsafely state _k v fp
+        | Just Refl <- eqT @k @GetModificationTime = pure v
+        -- GetModificationTime depends on these rules, so avoid creating a cycle
+        | Just Refl <- eqT @k @AddWatchedFile = liftIO getFromStoreUnsafely
```

Comment:
How do these work? If `GetModificationTime` depends on these, how can we rely on the value in the store to be accurate?

Reply:
We can't get the current value because it doesn't exist yet, but it doesn't matter too much because these rules do not produce diagnostics

Reply:
Maybe we can return `Nothing` then?

Reply:
Yes, that's another option. I'm not sure what the implications would be, other than no position mapping information.